packages/gcp/data_stream/audit/elasticsearch/ingest_pipeline/default.yml

```
@@ -105,6 +105,118 @@ processors:
         API_VERSION: (v\d+([a-z]+)?(\d+)?)
         RESOURCE_TYPE: ([a-z]+((\.[a-z0-9]+)+)?)
       ignore_missing: true
+  - script:
+      description: Appends any relevant entity to `related.entity` for all events
+      lang: painless
+      on_failure:
+        - set:
+            description: Adds error reason to the document
+            field: error.message
+            value: "{{{ _ingest.on_failure_message }}}"
+      source: |
+        void addFields(Set entities, String[] fields) {
+            for (String field : fields) {
+              addField(entities, field);
+            }
+        }
+
+        void addField(Set entities, String fieldName) {
+          addValue(entities, field(fieldName).get(null));
+        }
+
+        boolean addValues(Set entities, String[] values) {
+            boolean addedAll = true;
+            for (String value : values) {
+              addedAll = addedAll && addValue(entities, value);
+            }
+
+            return addedAll;
+        }
+
+        boolean addValue(Set entities, String value) {
+          if (value == null || value == "") {
+            return false;
+          }
+
+         return entities.add(value);
+        }
+
+        boolean isKubernetes = false;
+        if (ctx.json?.resource?.type != null) {
+          String resType = ctx.json.resource.type;
+          if (resType == "k8s_cluster" || resType == "gke_cluster" || resType == "kubernetes") {
+            isKubernetes = true;
+          }
+        }
+
+        // Using tree set to ensure a sorting is kept (testing purposes)
+        TreeSet entities = new TreeSet();
+
+        addField(entities, "json.proyoPayload.request.parent");
+        if (!isKubernetes) {
+            addFields(entities, new String[]{
+              "json.protoPayload.resourceName",
+              "json.protoPayload.response.user"
+            });
+        }
+
+
+        HashMap authInfo = field("json.protoPayload.authenticationInfo").get(new HashMap());
+        if (!isKubernetes) {
+          addValue(entities, authInfo.principalEmail);
+        }
+        addValues(entities, new String[]{
+          authInfo.principalSubject,
+          authInfo.serviceAccountKeyName
+        });
+        if (authInfo.serviceAccountDelegationInfo != null) {
+          if (!(authInfo.serviceAccountDelegationInfo instanceof List)) {
+            authInfo.serviceAccountDelegationInfo.get(new ArrayList()).stream().forEach(
+              delegInfo -> {
+                addValues(entities, new String[]{
+                  delegInfo.principalSubject,
+                  delegInfo.firstPartyPrincipal.principalEmail,
```

Have you tested with `firstPartyPrincipal` being null? If that's a possibility, I think this could break.

Painless has a nice syntax sugar for null accessing like Kotlin. you could do
```suggestion
                  delegInfo.firstPartyPrincipal?.principalEmail,
```

---

of course that applies for other non null checked chains